ipv8/configuration_pydantic.py

```diff
@@ -0,0 +1,402 @@
+        return v
+
+
+class RandomWalk(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "RandomWalk"
+        if "init" not in kwargs:
+            kwargs["init"] = {"timeout": 3.0}
+        super().__init__(**kwargs)
+
+
+class EdgeWalk(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "EdgeWalk"
+        super().__init__(**kwargs)
+
+
+class RandomChurn(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "RandomChurn"
+        if "peers" not in kwargs:
+            kwargs["peers"] = -1
+        if "init" not in kwargs:
+            kwargs["init"] = {'sample_size': 8, 'ping_interval': 10.0, 'inactive_time': 27.5, 'drop_time': 57.5}
+        super().__init__(**kwargs)
+
+
+class PeriodicSimilarity(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "PeriodicSimilarity"
+        if "peers" not in kwargs:
+            kwargs["peers"] = -1
+        super().__init__(**kwargs)
+
+
+class PingChurn(Walker):
+    def __init__(self, **kwargs):
+        kwargs["strategy"] = "PingChurn"
+        if "peers" not in kwargs:
+            kwargs["peers"] = -1
+        super().__init__(**kwargs)
+
+
+class Overlay(BaseConfigurationClass):
+    cls: str = Field(alias='class')
+    key: str = default["keys"][0]["alias"]
+    walkers: List[Walker] = [RandomWalk()]
+    bootstrappers: List[Bootstrapper] = [DispersyBootstrapper(init=DispersyBootstrapperInit())]
+    initialize: dict = Field(default_factory=dict)
+    on_start: List[tuple] = []
+
+
+class IPv8Configuration(BaseConfigurationClass):
+    """
+    Main pydantic IPv8 configuration model, can be fed directly into the ``IPv8`` class constructor.
+    """
+    interfaces: List[Interface] = [Interface()]
+    key_aliases: List[Key] = Field([FileKey()], alias="keys")
+    logger: Logger = Logger()
+    working_directory: str = default["working_directory"]
+    walker_interval: float = default["walker_interval"]
+    overlays: List[Overlay] = [
+        Overlay(cls="DiscoveryCommunity", walkers=[RandomWalk(), RandomChurn(), PeriodicSimilarity()]),
+        Overlay(cls="HiddenTunnelCommunity", initialize=default["overlays"][1]["initialize"],
+                on_start=default["overlays"][1]["on_start"]),
+        Overlay(cls="DHTDiscoveryCommunity", walkers=[RandomWalk(), PingChurn()])
+    ]
+
+    @validator("interfaces", always=True)
+    @classmethod
+    def validate_interfaces(cls, v):
+        it_names = [spec.interface for spec in v]
+        assert len(v) == len(set(it_names)), f"Duplicate interface names specified: {', '.join(it_names)}!"
+        return v
+
+
+def format_schema_recursive(*base_models: Type[BaseModel]) -> dict:
+    """
+    Create a schema (``dict``) describing a BaseModel class.
+
+    |
+    For example, the following pretty-prints the schema of ``IPv8Configuration``:
+
+    ::
+
+        from json import dumps
+        print(dumps(format_schema_recursive(IPv8Configuration), indent=4))
```

Comment:
With the goal of offering a runtime reference/reminder, I have created this method (instead of using the default) for two reasons:

1. Reducing the output size.
2. Increasing abstraction.

The output size difference is substantial. The default `schema_json` outputs 797 lines, while my method provides the necessary information in 49 lines. On my screen resolution and with my editor, I can see 52 lines at once. Therefore, this is the difference between reading through 16 pages of text or having all of the info at once.

The defaults and general abstraction level offered by `schema_json` are overly complex for the goal of capturing the essential schema. For example, lets take the `Overlay` and its `initialize` and `on_start` fields. In `schema_json` they are given as follows (nested 4 levels deep!):

```json
                "initialize": {
                    "title": "Initialize",
                    "type": "object"
                },
                "on_start": {
                    "title": "On Start",
                    "default": [],
                    "type": "array",
                    "items": {
                        "type": "array",
                        "items": {}
                    }
                }
```

My output is (disclaimer: in my opinion) equally expressive with regard to the schema but with less nesting (2 levels less) and lesser cognitive complexity:

```json
        "initialize": {},
        "on_start": []
```

(note: the above should actually be `"on_start": [[]]`, this is a bug)